### PR TITLE
Slider History MultiHashFilter implementation

### PR DESCRIPTION
**Pull request by andrewconner**
**Do not merge yet**

For Jared's review, to make sure this is what he needs. @2is10

### CODE REVIEW

File: shoebox/app/com/keepit/model/SliderHistory.scala [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import scala.collection.mutable
+import com.keepit.common.logging.Logging
+import play.api.libs.json._
+import com.keepit.common.cache.{FortyTwoCache, FortyTwoCachePlugin, Key}
+import com.keepit.serializer.SliderHistoryBinarySerializer
+import akka.util.duration._
+import com.keepit.search.MultiHashFilter
+
+case class SliderHistory (
+                             id: Option[Id[SliderHistory]] = None,
```
Comment (gmethvin) [inline shoebox/app/com/keepit/model/SliderHistory.scala]: Not sure what our convention is for indentation here but I prefer either 2 or 4 spaces.

If you're using IntelliJ you can turn off "align when multiline" and similar settings in your code style settings.

File: shoebox/app/com/keepit/model/SliderHistory.scala [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+object SliderHistoryTracker {
+  def apply(filterSize: Int, numHashFuncs: Int, minHits: Int) = {
+    new SliderHistoryTracker(filterSize, numHashFuncs, minHits)
+  }
+}
+
+class SliderHistoryTracker(tableSize: Int, numHashFuncs: Int, minHits: Int) {
+
+  def add(userId: Id[User], uriId: Id[NormalizedURI]) = {
```
Comment (gmethvin) [inline shoebox/app/com/keepit/model/SliderHistory.scala]: Please explicitly declare your return types on your public methods.

File: packrat/scripts/slider.js [hunk 3]
```diff
@@ -152,7 +152,7 @@ slider = function() {
             api.log("No need to inject, it's already here!");
           } else {
             drawKeepItHover(o.session, o.friends, o.numComments, o.numMessages, template, callback);
-            logEvent("slider", "sliderShown", {trigger: trigger, onPageMs: String(lastShownAt - t0), url: location.href});
+            logEvent("slider", "sliderShown", {trigger: trigger, url: document.location, onPageMs: String(lastShownAt - t0), url: location.href});
```
Comment (2is10) [inline packrat/scripts/slider.js]: Uh, there's already a `url:` in this event object. (I added it last week.)

File: shoebox/app/com/keepit/common/analytics/EventListener.scala [hunk 4]
```diff
@@ -77,6 +77,22 @@ class UsefulPageListener extends EventListenerPlugin {
...
 
+class SliderShownListener extends EventListenerPlugin {
+  private lazy val sliderHistoryTracker = inject[SliderHistoryTracker]
+
+  def onEvent: PartialFunction[Event,Unit] = {
+    case Event(_, UserEventMetadata(EventFamilies.SLIDER, "sliderShown", externalUser, _, experiments, metaData, _), _, _) =>
+      val (user, url, normUrl) = inject[DBConnection].readOnly {implicit s =>
+        val user = inject[UserRepo].get(externalUser)
+        val url = (metaData \ "url").asOpt[String].getOrElse("")
+        val normUrl = inject[NormalizedURIRepo].getByNormalizedUrl(URINormalizer.normalize(url))
```
Comment (2is10) [inline shoebox/app/com/keepit/common/analytics/EventListener.scala]: The `URINormalizer.normalize` call is redundant. Yes, `getByNormalizedUrl` would be more accurately named `getByNotYetNormalizedUrl` or perhaps just `getByUrl`.

File: shoebox/app/com/keepit/common/analytics/EventListener.scala [hunk 5]
```diff
@@ -77,6 +77,22 @@ class UsefulPageListener extends EventListenerPlugin {
...
+        (user, url, normUrl)
+      }
+      // handle UsefulPageListener
```
Comment (2is10) [inline shoebox/app/com/keepit/common/analytics/EventListener.scala]: what does this comment mean?

File: shoebox/app/com/keepit/model/SliderHistory.scala [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    val sliderHistoryRepo = inject[SliderHistoryRepo]
```
Comment (2is10) [inline shoebox/app/com/keepit/model/SliderHistory.scala]: unused

File: shoebox/app/com/keepit/model/SliderHistory.scala [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+  def getMultiHashFilter(userId: Id[User]) = {
+    val sliderHistoryRepo = inject[SliderHistoryRepo]
+    inject[DBConnection].readOnly { implicit session =>
+      sliderHistoryRepo.getByUserId(userId) match {
+        case Some(sliderHistory) =>
+          new MultiHashFilter(sliderHistory.tableSize, sliderHistory.filter, sliderHistory.numHashFuncs, sliderHistory.minHits)
+        case None =>
+          val filter = MultiHashFilter(tableSize, numHashFuncs, minHits)
+          filter
```
Comment (2is10) [inline shoebox/app/com/keepit/model/SliderHistory.scala]: how sweet of you to name this little guy

File: shoebox/app/com/keepit/model/SliderHistory.scala [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    }
+  }
+
+  def getMultiHashFilter(userId: Id[User]) = {
+    val sliderHistoryRepo = inject[SliderHistoryRepo]
+    inject[DBConnection].readOnly { implicit session =>
+      sliderHistoryRepo.getByUserId(userId) match {
+        case Some(sliderHistory) =>
+          new MultiHashFilter(sliderHistory.tableSize, sliderHistory.filter, sliderHistory.numHashFuncs, sliderHistory.minHits)
+        case None =>
```
Comment (gmethvin) [inline shoebox/app/com/keepit/model/SliderHistory.scala:124]: I noticed you used `new` here and not below. We should probably be consistent.

**Comment by 2is10** [comment]
Looks good, thanks! Feel free to merge when comments are addressed.

File: shoebox/app/com/keepit/model/SliderHistory.scala (hunk 8 above)
Comment (andrewconner) [inline shoebox/app/com/keepit/model/SliderHistory.scala:124]: It's not a case class, and the other is using the singleton's factory.

**Pull request merged**